src/reducers/entities/posts.js

```diff
@@ -1,530 +1,935 @@
+
+        const channelId = action.data.id;
+
+        let postDeleted = false;
+
+        // Remove any posts in the deleted channel
+        const nextState = {...state};
+        for (const post of Object.values(state)) {
+            if (post.channel_id === channelId) {
+                Reflect.deleteProperty(nextState, post.id);
+                postDeleted = true;
+            }
+        }
+
+        if (!postDeleted) {
+            // Nothing changed
+            return state;
+        }
+
+        return nextState;
+    }
+
+    case UserTypes.LOGOUT_SUCCESS:
+        return {};
+    default:
+        return state;
+    }
 }
 
-function handleRemovePendingPost(posts = {}, postsInChannel = {}, postsInThread = {}, action) {
-    const pendingPostId = action.data.id;
-    const channelId = action.data.channel_id;
-    const pendingPost = posts[pendingPostId];
+export function handlePendingPosts(state = [], action) {
+    switch (action.type) {
+    case PostTypes.RECEIVED_NEW_POST: {
+        const post = action.data;
 
-    const nextPosts = {
-        ...posts,
-    };
+        if (!post.pending_post_id) {
+            // This is not a pending post
+            return state;
+        }
 
-    Reflect.deleteProperty(nextPosts, pendingPostId);
+        const index = state.indexOf(post.pending_post_id);
 
-    let nextPostsInChannel = postsInChannel;
+        if (index !== -1) {
+            // An entry already exists for this post
+            return state;
+        }
 
-    // Only change postsInChannel if the order of the posts needs to change
-    if (!postsInChannel[channelId] || postsInChannel[channelId].includes(pendingPostId)) {
-        // If we don't already have the post, assume it's the most recent one
-        const postsForChannel = postsInChannel[channelId] || [];
+        // Add the new pending post ID
+        const nextState = [...state];
+        nextState.push(post.pending_post_id);
 
-        nextPostsInChannel = {...postsInChannel};
-        nextPostsInChannel[channelId] = postsForChannel.filter((postId) => postId !== pendingPostId);
+        return nextState;
     }
+    case PostTypes.POST_REMOVED: {
+        const post = action.data;
+
+        const index = state.indexOf(post.id);
+
+        if (index === -1) {
+            // There's nothing to remove
+            return state;
+        }
 
-    let nextPostsInThread = postsInThread;
-    if (pendingPost.root_id && (!postsInThread[pendingPost.root_id] || postsInThread[pendingPost.root_id].includes(pendingPostId))) {
-        const postsForThread = postsInThread[pendingPost.root_id] || [];
+        // The post has been removed, so remove the entry for it
+        const nextState = [...state];
+        nextState.splice(index, 1);
 
-        nextPostsInThread = {...postsInThread};
-        nextPostsInThread[pendingPost.root_id] = postsForThread.filter((postId) => postId !== pendingPostId);
+        return nextState;
     }
+    case PostTypes.RECEIVED_POST: {
```

Comment:
The comment I added to `receivedPost` explains this a bit more, but this happens after the current user has created a post. It may have worked slightly differently before, but with my changes, the creating a post will look like:
1. The pending post is created with a `receivedNewPost`
2. The request is sent to the server
3. The server responds and a `receivedPost` is dispatched which removes the pending post.